### PR TITLE
Fix the runtime-version choice.

### PR DESCRIPTION
**Pull request by RGafiyatullin**
Adresses: #9997 

When called in context of a block, the runtime-version, defined before that block's computation, is used; with a little exception for the cases when the context is the genesis-block. ([example for such a call](https://github.com/paritytech/substrate/blob/de8b34c814089e9561cbc1cdb4399bfc91d4cd46/client/finality-grandpa/src/lib.rs#L511))

When running in a constrained pruning mode (i.e. when a limited quantity of state-db snapshots is available) such an invocation will be available for all the non-pruned blocks except for the last one.

### CODE REVIEW

**Comment by skunert** [comment]
Basically this is Solution #3 from the issue. I am not sure I understand this comment:

> When running in a constrained pruning mode (i.e. when a limited quantity of state-db snapshots is available) such an invocation will be available for all the non-pruned blocks except for the last one.

If I call the runtime on some non-pruned block, why are we guaranteed to have the runtime from the previous block available and not pruned?

**Comment by wigy-opensource-developer** [comment]
> If I call the runtime on some non-pruned block, why are we guaranteed to have the runtime from the previous block available and not pruned?

There is [no guarantee](https://github.com/paritytech/substrate/pull/11209/files#diff-150ac465651e61379e693bed027607afccac194be6c5e2957c4ceb83cb259b02R128) for that at all.

**Comment by skunert** [comment]
> > If I call the runtime on some non-pruned block, why are we guaranteed to have the runtime from the previous block available and not pruned?
> 
> There is [no guarantee](https://github.com/paritytech/substrate/pull/11209/files#diff-150ac465651e61379e693bed027607afccac194be6c5e2957c4ceb83cb259b02R128) for that at all.

I think I misinterpreted the comment. I read "last one" as "best block", but it is the opposite, the first available non-pruned block. Got it 👍 . So we would accept that for these block you just can not call.

File: client/state-db/src/lib.rs [hunk 1]
```diff
@@ -352,7 +352,21 @@ impl<BlockHash: Hash + MallocSizeOf, Key: Hash + MallocSizeOf> StateDbSync<Block
 			(&mut self.pruning, &self.mode)
 		{
 			loop {
-				if pruning.window_size() <= constraints.max_blocks.unwrap_or(0) as u64 {
+				// `constraitns.max_blocks` is Some(N) — only the last N states should be available,
```
Comment (koute) [inline client/state-db/src/lib.rs:355]: `constraitns` - typo

**Comment by RGafiyatullin** [comment]
## Disclaimer

I might have slightly oversimplified the things, but nonetheless the following is how I understand it.

@skunert , @wigy-opensource-developer , please feel free to correct me if I am criminally wrong.

## Preface

When a block is computed — it yields:
- its hash;
- the State (conceptually — the State-trie, technically — the State-trie's root node's hash).

Once a block is canonicalised, the StateDb's "Pruning RefWindow" is notified.
The "Pruning RefWindow" appends that Block to the queue (aptly named `death_rows`).
The overall length of `death_rows` is kept under the specified `max_blocks`.

That as I understand it gives us "sort of guarantee" (a bit oxymoronic, hence the quotes) that 
- the states produced by the computation of non-canonicalised blocks are not pruned;
- the states produced by the computation of the `N` last canonicalised blocks  are not pruned.

The corner case here is when the `death_rows` queue is in fact shorter than the `max_blocks`:
- the whole history is shorter (the whole chain has just been started);
- the node has just restarted with the value for `max_blocks` larger than it used to run before the restart.

## Option # 3

There is an existing API `sc-client-api::Backend<B>::state_at(&self, block_id: BlockId<B>)`, it returns the State that is available as the result of the computation of block `block_id`.

So in order to 
a) compute that block again (`CallExecutor<B>::prove_execution(...)`);
b) query the data of that state (`CallExecutor<B>::call(...)`);
c) ... erm... (`CallExecutor::contextual_call(...)` — please help me out with this one, what is that?)

we would need to execute the version of the runtime available just before that block was computed, i.e. `state_at(block_id - 1).get(WellKnownKeys::CODE)`.

## What states can be queried?

In essence all states to which both of the following is true:
a) the state with the data is not pruned;
b) the state with the runtime is not pruned.

That is (for all `max_blocks > 0`):
- for every non-canonicalised block;
- for almost every non-pruned block (except for the oldest one).

**Comment by bkchr** [comment]
Disclaimer, I have written this before @RGafiyatullin and @andresilva commented, so this may contains duplicate information. Just posting it for posterity. 

I'm personally not convinced that using the wasm blob from the parent block is the best way to move forward (yes I know that I proposed it initially). When we do it that way we will need to change the pruning to keep one more block in a hidden way inside the unpruned blocks to always have the parent wasm blob around.
We also will need to "tag" the block production and block import to use the wasm blob from block X and not X - 1. Otherwise we would use the wrong wasm blob. Imagine we applied a runtime upgrade at block X with a migration. This migration needs to run while building/importing block X + 1. Building/importing a block is always done on the state of the parent block from which we currently take the runtime blob as well. With your changes we would take the runtime blob from the parent block of the parent of the block we are trying to build/import. If we would always have done it this way, it would have been no problem. However, now we need to stay backwards compatible and this means when you applied a runtime upgrade at block X, it needs to be used in block X + 1. (I hope you get what I want to say, if not please ask)

**Comment by wigy-opensource-developer** [comment]
Correct me if I am mistaken, but we cannot stay completely backward compatible for each of the following cases:

1. block importing
2. RPC runtime calls
3. block authoring

I really like the simplicity of this solution opposed to the really overengineered idea I had that also would allow multi-block migrations. If the `dead_row` keeps the state of a few finalized blocks alive in storage, I really do not see the issue with this direction.

**Comment by wigy-opensource-developer** [comment]
https://github.com/paritytech/substrate/blob/8ce50951e4c353bdb711e0c01306b0dfc8f2e561/client/service/src/client/call_executor.rs#L130-L133

Also, @RGafiyatullin, I think André tried to get you navigate here using the previous block hash instead of just decreasing the block number. That way is fork-aware.

**Comment by bkchr** [comment]
> Correct me if I am mistaken, but we cannot stay completely backward compatible for each of the following cases:
> 
>     1. block importing
> 
>     2. RPC runtime calls
> 
>     3. block authoring

These are basically all kind of interactions with the runtime ;) 

We will need to stay backwards compatible anyway, because otherwise we can not import all the old blocks.

**Comment by RGafiyatullin** [comment]
[In that](https://github.com/paritytech/substrate/pull/11209#issuecomment-1098124008) if block `N - 1` invoked `system::set_code` (or has set the state-db value under the key ":code" directly), then this block's post-state (i.e. `backend.state_at(N - 1)`) contains the runtime to compute block `N`.

**Comment by RGafiyatullin** [comment]
@bkchr please disregard the above... I traced a little bit more, I misunderstood what the input argument `at: BlockId<B>` for CallExecutor's method meant.

**Comment by RGafiyatullin** [comment]
I found a critical mistake in the very base of the reasoning that stands behind that PR-draft, therefore I will close it.

I will draw several doodles reflecting my up-to-date vision of the problem and post it in #9997 .

**Comment by bkchr** [comment]
Ty @RGafiyatullin

**Pull request closed without merging**